src/test/util/setup_common.h

```diff
@@ -39,13 +40,42 @@ extern const std::function<std::vector<const char*>()> G_TEST_COMMAND_LINE_ARGUM
 /** Retrieve the unit test name. */
 extern const std::function<std::string()> G_TEST_GET_FULL_NAME;
 
-// Enable BOOST_CHECK_EQUAL for enum class types
 namespace std {
+// Enable BOOST_CHECK_EQUAL for enum class types
 template <typename T>
-std::ostream& operator<<(typename std::enable_if<std::is_enum<T>::value, std::ostream>::type& stream, const T& e)
+inline std::ostream& operator<<(typename std::enable_if<std::is_enum<T>::value, std::ostream>::type& stream, const T& e)
 {
     return stream << static_cast<typename std::underlying_type<T>::type>(e);
 }
+
+// Enable BOOST_CHECK_EQUAL for std::optional
+template <typename T>
+inline std::ostream& operator<<(std::ostream& os, const std::optional<T>& v)
+{
+    if (v) {
+        return os << v.value();
+    } else {
+        return os << "std::nullopt";
```

Comment:
This seemed like a simplification to me, what would be the advantage of that?

Reply:
Done, but I don't yet see what this adds exactly

Reply:
Being able to write `BOOST_CHECK_EQUALS(..., std::nullopt);`, in case one wanted to.

Reply:
but we're never writing optionals on the right side here, i.e.
```C++
BOOST_CHECK_EQUAL(uint256::FromUserHex(""), std::make_optional(uint256::ZERO));
```
can simply be either:
```C++
BOOST_CHECK_EQUAL(uint256::FromUserHex(""), uint256::ZERO);
```
or if empty:
```C++
BOOST_CHECK(!uint256::FromUserHex(""));
```

Reply:
Removed the `nullopt` overwrite and explained the usage in the comment, thanks for explaining your view.

Reply:
I think I agree that until we have use case (I can't think of one atm) to compare with `std::nullopt`, forcing the user to use `BOOST_CHECK` makes sense.

In commit e0a96acc82c2c918572a2c9b90431db5bca19174:
> Note that we're never writing optionals on the right side, e.g.
> BOOST_CHECK_EQUAL(uint256::FromUserHex(""), std::make_optional(uint256::ZERO));
since they can simply be either:
> BOOST_CHECK_EQUAL(uint256::FromUserHex(""), uint256::ZERO);
or if empty:
> BOOST_CHECK(!uint256::FromUserHex(""));

I think this can be dropped from the commit message? It's possible, and there are good use cases for putting an optional on the right side, e.g. [here](https://github.com/stickies-v/bitcoin/commit/1f2b0fa86db2ed65476b68417aa1bf4c26026a19#diff-fc0f86b6c3bb23b0e983bcf79d7546d1c9eaa15d6e4d8a7b03b5b85955f585abR682-R684)). You could add a note to the developer notes on this, but I don't think that's necessary either.

Reply:
Wouldn't
```C++
        BOOST_CHECK_EQUAL(from_hex.has_value(), parse_hash_v.has_value());
        if (!from_hex.has_value()) continue;
        BOOST_CHECK_EQUAL(from_hex.value(), parse_hash_v.value());
```
be simpler after this change as just directly comparing the two?
```C++
        BOOST_CHECK_EQUAL(from_hex, parse_hash_v);
```
?

Reply:
Yes, in which case we'd be putting an optional (variable) on the right hand side, which goes against what you're saying in the commit message. That's why I suggest removing it, it doesn't help with reviewing this commit imo. It only adds context on why it doesn't do something that was done in another commit (that's not in the tree).

Reply:
> I think this can be dropped from the commit message

Done, your right, the formulation wasn't the best